| Problem Type                   | Recommended Algorithms                                  |
| ------------------------------ | ------------------------------------------------------- |
| Binary classification          | Logistic Regression, Decision Tree, Random Forest, SVM  |
| Multi-class classification     | Random Forest, XGBoost, KNN, Neural Networks            |
| Regression (Continuous output) | Linear Regression, Ridge/Lasso, Random Forest Regressor |
| Clustering                     | K-Means, DBSCAN, Hierarchical                           |
| Text classification            | Naive Bayes, Logistic Regression, LSTM                  |
| Image classification           | CNN, SVM (simple)                                       |
| Time series prediction         | ARIMA, Prophet, LSTM                                    |


In [1]:
import pandas as pd

In [3]:
df=pd.read_csv('forest_cover.csv')

In [5]:
df

,Id,Elevation,Aspect,Slope,Horizontal_Distance_To_Hydrology,Vertical_Distance_To_Hydrology,Horizontal_Distance_To_Roadways,Hillshade_9am,Hillshade_Noon,Hillshade_3pm,...,Soil_Type32,Soil_Type33,Soil_Type34,Soil_Type35,Soil_Type36,Soil_Type37,Soil_Type38,Soil_Type39,Soil_Type40,Cover_Type
0,1,2596,51,3,258,0,510,221,232,148,...,0,0,0,0,0,0,0,0,0,5
1,2,2590,56,2,212,-6,390,220,235,151,...,0,0,0,0,0,0,0,0,0,5
2,3,2804,139,9,268,65,3180,234,238,135,...,0,0,0,0,0,0,0,0,0,2
3,4,2785,155,18,242,118,3090,238,238,122,...,0,0,0,0,0,0,0,0,0,2
4,5,2595,45,2,153,-1,391,220,234,150,...,0,0,0,0,0,0,0,0,0,5
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
15115,15116,2607,243,23,258,7,660,170,251,214,...,0,0,0,0,0,0,0,0,0,3
15116,15117,2603,121,19,633,195,618,249,221,91,...,0,0,0,0,0,0,0,0,0,3
15117,15118,2492,134,25,365,117,335,250,220,83,...,0,0,0,0,0,0,0,0,0,3
15118,15119,2487,167,28,218,101,242,229,237,119,...,0,0,0,0,0,0,0,0,0,3


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15120 entries, 0 to 15119
Data columns (total 56 columns):
 #   Column                              Non-Null Count  Dtype
---  ------                              --------------  -----
 0   Id                                  15120 non-null  int64
 1   Elevation                           15120 non-null  int64
 2   Aspect                              15120 non-null  int64
 3   Slope                               15120 non-null  int64
 4   Horizontal_Distance_To_Hydrology    15120 non-null  int64
 5   Vertical_Distance_To_Hydrology      15120 non-null  int64
 6   Horizontal_Distance_To_Roadways     15120 non-null  int64
 7   Hillshade_9am                       15120 non-null  int64
 8   Hillshade_Noon                      15120 non-null  int64
 9   Hillshade_3pm                       15120 non-null  int64
 10  Horizontal_Distance_To_Fire_Points  15120 non-null  int64
 11  Wilderness_Area1                    15120 non-null  int64
 12  Wild

In [9]:
X = df.drop('Cover_Type',axis=1)
y = df['Cover_Type']

In [11]:
from sklearn.model_selection import train_test_split

In [13]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

In [15]:
X_train.shape # 12096 training observations,55 features

(12096, 55)

In [17]:
y_train.shape

(12096,)

| X\_train row          | y\_train value |
| --------------------- | -------------- |
| `[2596, 51, 3, ...]`  | `5`            |
| `[2590, 56, 2, ...]`  | `5`            |
| `[2800, 180, 6, ...]` | `2`            |
| ...                   | ...            |
Pehli row ka forest cover type = 5 (Aspen)
Doosri row ka forest cover type = 5 (Aspen)
Teesri row ka forest cover type = 2 (Lodgepole Pine)
Machine learning ko bas ek label chahiye hota hai

In [21]:
X_test.shape

(3024, 55)

### Train Models

In [24]:
from sklearn.linear_model import LogisticRegression

In [26]:
from sklearn.metrics import accuracy_score

In [28]:
lg=LogisticRegression()
lg.fit(X_train,y_train)

C:\Users\iaman\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [34]:
ypred=lg.predict(X_test)
accuracy_score_detection=accuracy_score(y_test,ypred)
print(accuracy_score_detection)

0.3826058201058201


DecisionTreeClassifier ek supervised ML algorithm hai. Ye classification problems ke liye use hota hai.
Ye data ko decision tree structure mein todta hai — questions poochta hai jaise:
"Is elevation > 2500?" "Is slope < 5°?"
Aur finally predict karta hai ki output kis class mein jayega.

In [38]:
from sklearn.tree import DecisionTreeClassifier

In [40]:
dtc=DecisionTreeClassifier()

In [42]:
dtc.fit(X_train,y_train)

DecisionTreeClassifier()

In [44]:
y_pred=dtc.predict(X_test)

In [48]:
accuracy_score_detection=accuracy_score(y_test,y_pred)
print(accuracy_score_detection)

0.8111772486772487


In [50]:
y_pred

array([6, 6, 1, ..., 5, 4, 4], dtype=int64)

KNeighborsClassifier ek supervised ML algorithm hai. Ye K-Nearest Neighbors (KNN) algorithm use karta hai.
Ye decision tree ki tarah tree nahi banata — balki: Jab prediction chahiye hota hai, tab training data mein sabse nazdeek wale k examples (neighbors) ko dekhta hai.Uske basis par vote karta hai: majority class kya hai.

In [52]:
from sklearn.neighbors import KNeighborsClassifier

In [54]:
knc=KNeighborsClassifier()
knc.fit(X_train,y_train)
y2_pred=knc.predict(X_test)
accuracy_score_detection=accuracy_score(y_test,y2_pred)
print(accuracy_score_detection)

0.8128306878306878


RandomForestClassifier ek supervised ensemble learning method hai. Ye Decision Trees ka forest banata hai.
Matlab: Ye ek tree nahi — bahut saare decision trees banata hai. Har tree training data ka thoda random part use karke train hota hai.
Sab trees milke vote karte hain final prediction ke liye.

In [59]:
from sklearn.ensemble import RandomForestClassifier

In [61]:
rfc=RandomForestClassifier()
rfc.fit(X_train,y_train)
y3_pred=rfc.predict(X_test)
accuracy_score_detection=accuracy_score(y_test,y3_pred)
print(accuracy_score_detection)

0.8839285714285714


In [63]:
y3_pred

array([6, 3, 1, ..., 5, 4, 4], dtype=int64)

### Predictoin system

iloc = integer location . Ye row/column ko index number se select karta hai.
Jaise: df.iloc[0] → pehli row . df.iloc[10] → 11th row

In [68]:
df.iloc[10].values

array([  11, 2612,  201,    4,  180,   51,  735,  218,  243,  161, 6222,
          1,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    1,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          5], dtype=int64)

In [70]:
input_values = (101, 2998,   45,    8,  351,   16, 5842,  223,  222,  134, 3721,
          1,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    1,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0)

In [72]:
import numpy as np

In [74]:
features=np.array([input_values])

In [76]:
features

array([[ 101, 2998,   45,    8,  351,   16, 5842,  223,  222,  134, 3721,
           1,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    1,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0]])

In [78]:
rfc.predict(features)

C:\Users\iaman\anaconda3\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


array([2], dtype=int64)

In [80]:
#Jab tumne RandomForestClassifier ko train kiya, toh tumne X_train diya: X_train ek DataFrame tha → isme feature names (column names) the: ['Elevation', 'Aspect', ...].
#Jab tum predict kar rahe ho: Tumne features diya jo DataFrame nahi hai — balki Numpy array hai ya list hai.
#sklearn tumhe warn kar raha hai: "Bhai! Train karte waqt mujhe feature names diye the, par predict karte waqt mujhe feature names nahi mile!"

In [84]:
input_df = pd.DataFrame(features, columns=X_train.columns)

# Predict karo:
result = rfc.predict(input_df)
print(result)

[2]


Ye trained Random Forest Classifier (rfc) ko file mein save karti hai — taaki baad mein dobara train na karna pade.
Isko model serialization kehte hain.pickle ek Python module hai.
Iska kaam hai Python object ko bytes mein convert karna (serialize karna) ya wapas object banana (deserialize karna).
Python ke built-in objects: jaise list, dict, model — sab pickle se file mein store ho sakte hain.

Ye trained Random Forest Classifier (rfc) ko file mein save karti hai — taaki baad mein dobara train na karna pade.
Isko model serialization kehte hain.

dump function ka kaam: Python object ko serialize karke ek file mein save karo
pickle.dump(object, file_object) . 'rfc.pkl' → Ye output file ka naam hai. .pkl ka matlab: Pickle file.
'wb' → Ye mode hai: w → write (file mein likhna). b → binary mode (kyunki pickle binary data banata hai).

In [92]:
import pickle
pickle.dump(rfc,open('rfc.pkl','wb'))